cloudsync/providers/file.py

```diff
@@ -0,0 +1,504 @@
+import os
+import sys
+import glob
+import time
+import re
+import tempfile
+import logging
+import shutil
+import collections
+import threading
+from hashlib import blake2b
+import typing
+from dataclasses import dataclass
+
+from watchdog import events as watchdog_events
+from watchdog.observers import Observer as watchdog_observer
+
+from cloudsync.event import Event
+from cloudsync.provider import Provider
+from cloudsync.registry import register_provider
+from cloudsync.types import OInfo, OType, DirInfo
+import cloudsync.exceptions as ex
+
+
+def is_osx():
+    return sys.platform == "darwin"
+
+
+if is_osx():
+    from Foundation import NSURL  # pylint: disable=import-error,no-name-in-module
+
+
+logging.getLogger("watchdog").setLevel(logging.INFO)
+log = logging.getLogger(__name__)
+
+
+def get_hash(dat):
+    """Returns a hash of it's argument which can be a bytes or filelike object"""
+    # 32-byte blake optimized for 64 bit chips
+    if not hasattr(dat, "read"):
+        return blake2b(dat, digest_size=32).digest()
+    blocksize = 4096
+    d = blake2b(digest_size=32)
+    block = dat.read(blocksize)
+    while block:
+        d.update(block)
+        block = dat.read(blocksize)
+    return d.digest()
+
+
+def detect_case_sensitive(tmpdir=None):
+    """Returns true if temp direcory specified is case sensitive, or gettempdir() if unspecified."""
+    if not tmpdir:
+        tmpdir = tempfile.gettempdir()
+    f1 = os.path.join(tmpdir, "tmp." + os.urandom(16).hex())
+    with open(f1, "w") as f:
+        f.write("x")
+    try:
+        f2 = f1.upper()
+        if os.path.exists(f2):
+            return False
+    finally:
+        os.unlink(f1)
+    return True
+
+
+def casedpath(path):
+    """Fixes the case of a file to the name on disk.
+
+    Works only for a path that exists
+    """
+    if is_osx():
+        url = NSURL.fileURLWithPath_(path)  # will be None if path doesn't exist
+        if not url:
+            return path
+        return url.fileReferenceURL().path()
+
+    r = glob.glob(re.sub(r'([^:/\\])(?=[/\\]|$)', r'[\1]', path))
+    return r[0] if r else path
+
+
+def canonicalize_fpath(case_sensitive: bool, full_path: str) -> str:
+    """Fixes the case of a path - parent folder must exist."""
+    if not case_sensitive:
+        # not needed for case sensitive installations
+        return full_path
+
+    if not os.path.exists(full_path):
+        # casedpath function doesn't work if the path doesn't exist
+        fdir, fname = os.path.split(full_path)
+        cp = casedpath(fdir)
+        fp: typing.Optional[str]
+        if cp:
```

Comment:
is this how casedpath actually works? returns false if it doesn't exist? It looks like it just returns fdir right back out if it doesn't exist. Perhaps check `os.path.exists` for fdir explicitly instead of using cp this way

Reply:
fixed